File: controllers/households.js
const { AppUser, Household, HouseholdMember, ExpenseCategory, IncomeCategory } = require('../models');

class HouseholdsController {
  /**
   * List all households (paginated), including member count and member list
   */
  async list(ctx) {
    try {
      const page = parseInt(ctx.query.page) || 1;
      const limit = parseInt(ctx.query.limit) || 20;
      const offset = (page - 1) * limit;

      const { count, rows: households } = await Household.findAndCountAll({
        include: [
          {
            model: HouseholdMember,
            include: [
              {
                model: AppUser,
                attributes: ['id', 'username']
              }
            ]
          }
        ],
        limit,
        offset,
        order: [['createdAt', 'DESC']]
      });

      const pages = Math.ceil(count / limit);

      ctx.body = {
        success: true,
        households: households.map(h => {
          const json = h.toJSON();
          json.memberCount = json.HouseholdMembers ? json.HouseholdMembers.length : 0;
          return json;
        }),
        pagination: {
          total: count,
          page,
          limit,
          pages
        }
      };
    } catch (error) {
      console.error('List households error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to list households' };
    }
  }

  /**
   * Get single household by id, including full member list, expense and income categories
   */
  async get(ctx) {
    try {
      const { id } = ctx.params;

      const household = await Household.findByPk(id, {
        include: [
          {
            model: HouseholdMember,
            include: [
              {
                model: AppUser,
                attributes: ['id', 'username', 'isActive', 'lastLogin', 'createdAt']
              }
            ]
          },
          {
            model: ExpenseCategory
          },
          {
            model: IncomeCategory
          }
        ]
      });

      if (!household) {
        ctx.status = 404;
        ctx.body = { error: 'Household not found' };
        return;
      }

      ctx.body = {
        success: true,
        household: household.toJSON()
      };
    } catch (error) {
      console.error('Get household error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to get household' };
    }
  }

  /**
   * Create a new household
   */
  async create(ctx) {
    try {
      const { name, description } = ctx.request.body;

      if (!name) {
        ctx.status = 400;
        ctx.body = { error: 'Name is required' };
        return;
      }

      const household = await Household.create({ name, description });

      ctx.status = 201;
      ctx.body = {
        success: true,
        household: household.toJSON()
      };
    } catch (error) {
      console.error('Create household error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to create household' };
    }
  }

  /**
   * Update a household's name and/or description
   */
  async update(ctx) {
    try {
      const { id } = ctx.params;
      const { name, description } = ctx.request.body;

      const household = await Household.findByPk(id);

      if (!household) {
        ctx.status = 404;
        ctx.body = { error: 'Household not found' };
        return;
      }

      await household.update({
        name: name !== undefined ? name : household.name,
        description: description !== undefined ? description : household.description
      });

      ctx.body = {
        success: true,
        household: household.toJSON()
      };
    } catch (error) {
      console.error('Update household error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to update household' };
    }
  }

  /**
   * Delete a household
   */
  async delete(ctx) {
    try {
      const { id } = ctx.params;

      const household = await Household.findByPk(id);

      if (!household) {
        ctx.status = 404;
        ctx.body = { error: 'Household not found' };
        return;
      }

      await household.destroy();

      ctx.body = { success: true };
    } catch (error) {
      console.error('Delete household error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to delete household' };
    }
  }

  /**
   * Add a member to a household
   */
  async addMember(ctx) {
    try {
      const { id } = ctx.params;
      const { appUserId, role } = ctx.request.body;

      if (!appUserId) {
        ctx.status = 400;
        ctx.body = { error: 'appUserId is required' };
        return;
      }

      const household = await Household.findByPk(id);

      if (!household) {
        ctx.status = 404;
        ctx.body = { error: 'Household not found' };
        return;
      }

      const appUser = await AppUser.findByPk(appUserId);

      if (!appUser) {
        ctx.status = 404;
        ctx.body = { error: 'App user not found' };
        return;
      }

      const existingMember = await HouseholdMember.findOne({
        where: { householdId: id, appUserId }
      });

      if (existingMember) {
        ctx.status = 409;
        ctx.body = { error: 'User is already a member of this household' };
        return;
      }

      await HouseholdMember.create({
        householdId: id,
        appUserId,
        role: role || 'member'
      });

      const members = await HouseholdMember.findAll({
        where: { householdId: id },
        include: [
          {
            model: AppUser,
            attributes: ['id', 'username', 'isActive']
          }
        ]
      });

      ctx.status = 201;
      ctx.body = {
        success: true,
        members: members.map(m => m.toJSON())
      };
    } catch (error) {
      console.error('Add household member error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to add member to household' };
    }
  }

  /**
   * Remove a member from a household
   */
  async removeMember(ctx) {
    try {
      const { id, appUserId } = ctx.params;

      const member = await HouseholdMember.findOne({
        where: { householdId: id, appUserId }
      });

      if (!member) {
        ctx.status = 404;
        ctx.body = { error: 'Household member not found' };
        return;
      }

      await member.destroy();

      ctx.body = { success: true };
    } catch (error) {
      console.error('Remove household member error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to remove member from household' };
    }
  }

  /**
   * Update a member's role within a household
   */
  async updateMemberRole(ctx) {
    try {
      const { id, appUserId } = ctx.params;
      const { role } = ctx.request.body;

      if (!role) {
        ctx.status = 400;
        ctx.body = { error: 'Role is required' };
        return;
      }

      const member = await HouseholdMember.findOne({
        where: { householdId: id, appUserId }
      });

      if (!member) {
        ctx.status = 404;
        ctx.body = { error: 'Household member not found' };
        return;
      }

      await member.update({ role });

      const updatedMember = await HouseholdMember.findOne({
        where: { householdId: id, appUserId },
        include: [
          {
            model: AppUser,
            attributes: ['id', 'username', 'isActive']
          }
        ]
      });

      ctx.body = {
        success: true,
        member: updatedMember.toJSON()
      };
    } catch (error) {
      console.error('Update household member role error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to update member role' };
    }
  }
}

module.exports = new HouseholdsController();

File: backend_go/internal/handlers/households.go
package handlers

import (
	"household-go/internal/database"
	"household-go/internal/models"
	"household-go/internal/utils"
	"log"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// HouseholdsHandler handles household CRUD and member management.
type HouseholdsHandler struct{}

func NewHouseholdsHandler() *HouseholdsHandler { return &HouseholdsHandler{} }

// List returns paginated households with member counts.
// GET /api/households?page=1&limit=20
func (h *HouseholdsHandler) List(c *gin.Context) {
	page := max(parseInt(c.DefaultQuery("page", "1")), 1)
	limit := parseInt(c.DefaultQuery("limit", "20"))
	if limit <= 0 || limit > 100 {
		limit = 20
	}
	offset := (page - 1) * limit

	var total int64
	var households []models.Household

	// Count and query in parallel — two separate DB calls.
	// GORM doesn't have a findAndCountAll() like Sequelize; you do it manually.
	database.DB.Model(&models.Household{}).Count(&total)
	database.DB.
		Preload("Members", func(db *gorm.DB) *gorm.DB {
			// Preload nested associations: Members → AppUser
			return db.Preload("AppUser", func(db *gorm.DB) *gorm.DB {
				return db.Select("id, username")
			})
		}).
		Order("created_at DESC").
		Limit(limit).
		Offset(offset).
		Find(&households)

	pages := int((total + int64(limit) - 1) / int64(limit)) // ceiling division

	utils.OK(c, gin.H{
		"success":    true,
		"households": households,
		"pagination": gin.H{
			"total": total,
			"page":  page,
			"limit": limit,
			"pages": pages,
		},
	})
}

// Get returns a single household with full member list and categories.
// GET /api/households/:id
func (h *HouseholdsHandler) Get(c *gin.Context) {
	id := c.Param("id")

	var household models.Household
	err := database.DB.
		Preload("Members", func(db *gorm.DB) *gorm.DB {
			return db.Preload("AppUser", func(db *gorm.DB) *gorm.DB {
				return db.Select("id, username, is_active, last_login, created_at")
			})
		}).
		Preload("ExpenseCategories").
		Preload("IncomeCategories").
		First(&household, id).Error

	if err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.NotFound(c, "Household not found")
		} else {
			utils.InternalError(c, "Failed to get household")
		}
		return
	}

	utils.OK(c, gin.H{"success": true, "household": household})
}

// Create creates a new household.
// POST /api/households
func (h *HouseholdsHandler) Create(c *gin.Context) {
	var body struct {
		Name        string  `json:"name"        binding:"required"`
		Description *string `json:"description"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Name is required")
		return
	}

	household := models.Household{Name: body.Name, Description: body.Description}
	if err := database.DB.Create(&household).Error; err != nil {
		log.Printf("Create household error: %v", err)
		utils.InternalError(c, "Failed to create household")
		return
	}

	utils.Created(c, gin.H{"success": true, "household": household})
}

// Update changes a household's name and/or description.
// PUT /api/households/:id
func (h *HouseholdsHandler) Update(c *gin.Context) {
	id := c.Param("id")

	var household models.Household
	if err := database.DB.First(&household, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.NotFound(c, "Household not found")
		} else {
			utils.InternalError(c, "Failed to update household")
		}
		return
	}

	var body struct {
		Name        *string `json:"name"`
		Description *string `json:"description"`
	}
	c.ShouldBindJSON(&body)

	updates := map[string]interface{}{}
	if body.Name != nil        { updates["name"] = *body.Name }
	if body.Description != nil { updates["description"] = body.Description }

	if len(updates) > 0 {
		database.DB.Model(&household).Updates(updates)
	}

	utils.OK(c, gin.H{"success": true, "household": household})
}

// Delete removes a household.
// DELETE /api/households/:id
func (h *HouseholdsHandler) Delete(c *gin.Context) {
	id := c.Param("id")

	var household models.Household
	if err := database.DB.First(&household, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.NotFound(c, "Household not found")
		} else {
			utils.InternalError(c, "Failed to delete household")
		}
		return
	}

	database.DB.Delete(&household)
	utils.OK(c, gin.H{"success": true})
}

// AddMember adds an AppUser to a household.
// POST /api/households/:id/members
func (h *HouseholdsHandler) AddMember(c *gin.Context) {
	id := c.Param("id")

	var body struct {
		AppUserID uint   `json:"appUserId" binding:"required"`
		Role      string `json:"role"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "appUserId is required")
		return
	}

	var household models.Household
	if err := database.DB.First(&household, id).Error; err != nil {
		utils.NotFound(c, "Household not found")
		return
	}

	var appUser models.AppUser
	if err := database.DB.First(&appUser, body.AppUserID).Error; err != nil {
		utils.NotFound(c, "App user not found")
		return
	}

	// Check for existing membership
	var existing models.HouseholdMember
	if database.DB.Where("household_id = ? AND app_user_id = ?", id, body.AppUserID).First(&existing).Error == nil {
		utils.Conflict(c, "User is already a member of this household")
		return
	}

	role := body.Role
	if role == "" {
		role = "member"
	}

	member := models.HouseholdMember{
		HouseholdID: household.ID,
		AppUserID:   body.AppUserID,
		Role:        role,
	}
	database.DB.Create(&member)

	var members []models.HouseholdMember
	database.DB.Where("household_id = ?", id).
		Preload("AppUser", func(db *gorm.DB) *gorm.DB {
			return db.Select("id, username, is_active")
		}).Find(&members)

	utils.Created(c, gin.H{"success": true, "members": members})
}

// RemoveMember removes an AppUser from a household.
// DELETE /api/households/:id/members/:appUserId
func (h *HouseholdsHandler) RemoveMember(c *gin.Context) {
	id := c.Param("id")
	appUserID := c.Param("appUserId")

	var member models.HouseholdMember
	if err := database.DB.Where("household_id = ? AND app_user_id = ?", id, appUserID).
		First(&member).Error; err != nil {
		utils.NotFound(c, "Household member not found")
		return
	}

	database.DB.Delete(&member)
	utils.OK(c, gin.H{"success": true})
}

// UpdateMemberRole changes a member's role in a household.
// PUT /api/households/:id/members/:appUserId
func (h *HouseholdsHandler) UpdateMemberRole(c *gin.Context) {
	id := c.Param("id")
	appUserID := c.Param("appUserId")

	var body struct {
		Role string `json:"role" binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Role is required")
		return
	}

	var member models.HouseholdMember
	if err := database.DB.Where("household_id = ? AND app_user_id = ?", id, appUserID).
		First(&member).Error; err != nil {
		utils.NotFound(c, "Household member not found")
		return
	}

	database.DB.Model(&member).Update("role", body.Role)

	database.DB.Where("household_id = ? AND app_user_id = ?", id, appUserID).
		Preload("AppUser", func(db *gorm.DB) *gorm.DB {
			return db.Select("id, username, is_active")
		}).First(&member)

	utils.OK(c, gin.H{"success": true, "member": member})
}

// max returns the larger of two ints — Go 1.21+ has built-in max(), but we define
// it here for clarity. In Go 1.21+, you can remove this and use the builtin.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
